### PR TITLE
DAOS-5195 object: Add dc_tx locking

### PR DESCRIPTION
**Pull request by liw**
dc_tx should be thread-safe. This patch adds a pthread_mutex_t tx_lock,
which is supposed to be held for short durations. Since almost all use
cases need to modify dc_tx, this lock is not a pthread_rwlock_t.

Signed-off-by: Li Wei <wei.g.li@intel.com>

### CODE REVIEW

**Comment by liw** [comment]
> This looks fine to me, although i have to admit it sounds a bit ugly to lock in one place then unlock in another. If there are any bugs here, it might be a little hard to debug this.

Agreed. I had a similar feeling when writing this patch and looking at the dc_tx references passed around.

**Comment by liw** [comment]
Rebuild 25 timeout: DAOS-5256.

File: src/object/obj_tx.c [hunk 1]
```diff
@@ -332,6 +344,8 @@ dc_tx_check_pmv_internal(daos_handle_t th, struct dc_tx **ptx)
 	if (tx == NULL)
 		return -DER_NO_HDL;
 
+	D_MUTEX_LOCK(&tx->tx_lock);
```
Comment (jolivier23) [inline src/object/obj_tx.c:347]: dc_tx_check_pmv is called from src/object/cli_obj.c.   Where does the lock get released in this case?

Comment (liw) [reply src/object/obj_tx.c:347]: dc_tx_check_pmv calls dc_tx_check_pmv_internal with a NULL ptx. The lock is, therefore, released... (See below.)

File: src/object/obj_tx.c [hunk 2]
```diff
@@ -349,10 +363,12 @@ dc_tx_check_pmv_internal(daos_handle_t th, struct dc_tx **ptx)
 	}
 
 out:
-	if (rc != 0 || ptx == NULL)
+	if (rc != 0 || ptx == NULL) {
+		D_MUTEX_UNLOCK(&tx->tx_lock);
```
Comment (liw) [inline src/object/obj_tx.c:367]: ... released here.

**Comment by liw** [comment]
> just doing request changes to get it out of my inbox til I get answer to question

@jolivier23, I've answered the question.

**Comment by Nasf-Fan** [comment]
@daos-stack/daos-gatekeeper , this patch will cause a lot of code conflict with my subsequent DTX patches, so if possible, please consider to force land it, then I can merge it with my DTX patches. Thanks!

**Pull request merged**